rpcs3/Emu/Io/usio.cpp

```diff
@@ -0,0 +1,452 @@
+// v406 USIO emulator
+// Responses may be specific to Taiko no Tatsujin
+
+#include "stdafx.h"
+#include "usio.h"
+#include "Emu/Cell/lv2/sys_usbd.h"
+#include "Input/pad_thread.h"
+#include "Emu/System.h"
+
+LOG_CHANNEL(usio_log);
+
+usb_device_usio::usb_device_usio()
+{
+	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE,
+		UsbDeviceDescriptor{
+			.bcdUSB             = 0x0110,
+			.bDeviceClass       = 0xff,
+			.bDeviceSubClass    = 0x00,
+			.bDeviceProtocol    = 0xff,
+			.bMaxPacketSize0    = 0x8,
+			.idVendor           = 0x0b9a,
+			.idProduct          = 0x0910,
+			.bcdDevice          = 0x0910,
+			.iManufacturer      = 0x01,
+			.iProduct           = 0x02,
+			.iSerialNumber      = 0x00,
+			.bNumConfigurations = 0x01});
+
+	auto& config0 = device.add_node(UsbDescriptorNode(USB_DESCRIPTOR_CONFIG,
+		UsbDeviceConfiguration{
+			.wTotalLength        = 39,
+			.bNumInterfaces      = 0x01,
+			.bConfigurationValue = 0x01,
+			.iConfiguration      = 0x00,
+			.bmAttributes        = 0xc0,
+			.bMaxPower           = 0x32 // ??? 100ma
+		}));
+
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_INTERFACE,
+		UsbDeviceInterface{
+			.bInterfaceNumber   = 0x00,
+			.bAlternateSetting  = 0x00,
+			.bNumEndpoints      = 0x03,
+			.bInterfaceClass    = 0x00,
+			.bInterfaceSubClass = 0x00,
+			.bInterfaceProtocol = 0x00,
+			.iInterface         = 0x00}));
+
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT,
+		UsbDeviceEndpoint{
+			.bEndpointAddress = 0x01,
+			.bmAttributes     = 0x02,
+			.wMaxPacketSize   = 0x0040,
+			.bInterval        = 0x00}));
+
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT,
+		UsbDeviceEndpoint{
+			.bEndpointAddress = 0x82,
+			.bmAttributes     = 0x02,
+			.wMaxPacketSize   = 0x0040,
+			.bInterval        = 0x00}));
+
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT,
+		UsbDeviceEndpoint{
+			.bEndpointAddress = 0x83,
+			.bmAttributes     = 0x03,
+			.wMaxPacketSize   = 0x0008,
+			.bInterval        = 16}));
+}
+
+usb_device_usio::~usb_device_usio()
+{
+}
+
+void usb_device_usio::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue, u16 wIndex, u16 wLength, u32 buf_size, u8* buf, UsbTransfer* transfer)
+{
+	transfer->fake = true;
+
+	// Control transfers are nearly instant
+	switch (bmRequestType)
+	{
+	default:
+		// Follow to default emulated handler
+		usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+		break;
+	}
+}
+
+void usb_device_usio::translate_input()
+{
+	const auto handler = pad::get_current_handler();
```

Comment:
There should be a mutex here to ensure that the pad data is not messed with while we're trying to fetch it.
The same seems to be missing in Turntable.cpp

```suggestion
	std::lock_guard lock(pad::g_pad_mutex);
	const auto handler = pad::get_current_handler();
```